rx/subject/replaysubject.py

```diff
@@ -0,0 +1,137 @@
+    interval: datetime
+    value: Any
+
+
+class ReplaySubject(Subject):
+    """Represents an object that is both an observable sequence as well
+    as an observer. Each notification is broadcasted to all subscribed
+    and future observers, subject to buffer trimming policies.
+    """
+
+    def __init__(self,
+                 buffer_size: int = None,
+                 window: typing.RelativeTime = None,
+                 scheduler: Optional[typing.Scheduler] = None
+                 ) -> None:
+        """Initializes a new instance of the ReplaySubject class with
+        the specified buffer size, window and scheduler.
+
+        Args:
+            buffer_size: [Optional] Maximum element count of the replay
+                buffer.
+            window [Optional]: Maximum time length of the replay buffer.
+            scheduler: [Optional] Scheduler the observers are invoked on.
+        """
+
+        super().__init__()
+        self.buffer_size = sys.maxsize if buffer_size is None else buffer_size
+        self.scheduler = scheduler or current_thread_scheduler
+        self.window = timedelta.max if window is None else self.scheduler.to_timedelta(window)
+        self.queue: List[QueueItem] = []
+
+    def _subscribe_core(self,
+                        observer: typing.Observer,
+                        scheduler: Optional[typing.Scheduler] = None
+                        ) -> typing.Disposable:
+        so = ScheduledObserver(self.scheduler, observer)
+        subscription = RemovableDisposable(self, so)
+
+        with self.lock:
+            self.check_disposed()
+            self._trim(self.scheduler.now)
+            self.observers.append(so)
+
+            for item in self.queue:
+                so.on_next(item.value)
+
+            if self.exception is not None:
+                so.on_error(self.exception)
+            elif self.is_stopped:
+                so.on_completed()
+
+        so.ensure_active()
+        return subscription
+
+    def _trim(self, now: datetime):
+        while len(self.queue) > self.buffer_size:
+            self.queue.pop(0)
+
+        while self.queue and (now - self.queue[0].interval) > self.window:
+            self.queue.pop(0)
+
+    def _on_next_core(self, value: Any) -> None:
+        """Notifies all subscribed observers with the value."""
+
+        with self.lock:
+            observers = self.observers.copy()
+            now = self.scheduler.now
+            self.queue.append(QueueItem(interval=now, value=value))
+            self._trim(now)
+
+        for observer in observers:
+            observer.on_next(value)
+
+        for observer in observers:
+            cast(ScheduledObserver, observer).ensure_active()
+
+    def _on_error_core(self, error: Exception) -> None:
+        """Notifies all subscribed observers with the exception."""
+
+        with self.lock:
+            observers = self.observers.copy()
+            self.observers.clear()
+            self.is_stopped = True
```

Comment:
`is_stopped` should already be set by `Observer.on_error`. Do we need to set it here as well? Same with `_on_competed_core`.

Reply:
Ah, good catch. No, that's not needed, I've amended the commit.

By the way, I believe the dropped coverage reported by coveralls is simply due to the fact that some code which was at 100% (and remains at 100%) became smaller -- and therefore the "gaps" in the other code become a slightly larger percentage of the whole.

Reply:
@dbrattli About these spurious dropped coverage reports, I think it looks bad on the list of master commits to have so many red crosses. I'm not not project owner so I cannot do this myself, but apparently you can set the threshold for this on 
https://coveralls.io/github/ReactiveX/RxPY/settings

I would suggest setting this "COVERAGE DECREASE THRESHOLD FOR FAILURE" at something like 0.1%. If that makes sense, could you change it? That seems like a safe margin to assume that anything below that is probably fine.

Note, we'll still see the percentages, I think, it's just that the build status would remain green for percentages less than the threshold.

Thanks!

Reply:
Thanks. Fixed!